lib/utils/url.sh

```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
+
+if [ -z "${BPKG_URL}" ]; then
+  BPKG_URL=1
+
+  ## Collection of shared bpkg functions
+
+  bpkg_resolve_uri_default() {
+    local git_remote=$1
+    local user=$2
+    local name=$3
+    local version=$4
+
+    if [[ "$git_remote" == https://gitlab*  ]]; then
+      echo "$user/$name/-/raw/$version"
+    else
+      echo "$user/$name/$version"
+    fi
+  }
+
+  bpkg_resolve_uri() {
+    [[ $(type -t bpkg_resolve_uri_pathspec) == function ]] && \
```

Comment:
Nit: I think something like this is more readable and removes the shellcheck warning:

```sh
if [[ $(type -t bpkg_resolve_uri_pathspec) == function ]]; then
	bpkg_resolve_uri_pathspec "$@" || bpkg_resolve_uri_default "$@"
else
  bpkg_resolve_uri_default "$@"
fi
```

It preserves the behavior where if `bpkg_resolve_uri_pathspec` fails, we then use our `bpkg_resolve_uri_default` function to resolve the uri (i think it's easy to miss that detail). But, should this be considered a bug? If I define a `bpkg_resolve_uri_pathspec` myself, I expect it to be the only function used to resolve URI, even if my function returns with a non-exit code. It might be better to keep the "string" blank as-is, or to print an error instead.

Reply:
the idea behind `bpkg_resolve_uri_pathspec` its not to be the only solver but a complimentary one to help everyone else solve their problem and stil let the heavy lifting to `bpkg` 

About the code suggestion is a very good one. I take it

Reply:
Fixed at https://github.com/bpkg/bpkg/pull/171/commits/be225d1e7c550d52d0841a02fe38dad815e651c8

Reply:
Thanks for the update. I do still think that both `bpkg_resolve_uri_pathspec` and `bpkg_resolve_uri_default` _shouldn't_ be invoked in the same codepath sequentially like that. I would be interested in hearing the other maintainers weigh in, but in any case, I'm not trying to stop this code from being merged as-is - it's a smaller detail.